jupyterhub/files/hub/jupyterhub_config.py

```diff
@@ -360,13 +365,17 @@ def camelCaseify(s):
     )
 
 for name, service in get_config("hub.services", {}).items():
-    # jupyterhub.services is a list of dicts, but
-    # in the helm chart it is a dict of dicts for easier merged-config
+    # c.JupyterHub.services is a list of dicts, but
+    # hub.services is a dict of dicts to make the config mergable
     service.setdefault("name", name)
-    # handle camelCase->snake_case of api_token
-    api_token = service.pop("apiToken", None)
+
+    # As the api_token could be exposed in hub.existingSecret, we need to read
+    # it it from there or fall back to the chart managed k8s Secret's value.
+    service.pop("apiToken", None)
+    api_token = get_secret_value(f"hub.services.{service['name']}.api_token", None)
```

Comment:
What happens if service.apiToken is defined in values as it normally is now?

I think perhaps what we want is this:

```suggestion
    api_token = service.pop("apiToken", None)
    api_token = get_secret_value(f"hub.services.{service['name']}.api_token", api_token)
```

so that we load it from values, and allow override from the secret.

Related: in .Values we expect camelCase `apiToken`, but in secret we expect snake_case `api_token`. Is that intentional?

Also related: is there a reason the custom secret values are not merged into the config so that `get_config` works with values from any source?

Reply:
> What happens if service.apiToken is defined in values as it normally is now?

It will be written to the chart managed k8s Secret, and then read by get_secret_value() which will favor any values coming from the mounted self managed k8s Secret if hub.existingSecret was set.

So, all good!

---

When rendering the self managed k8s secret i made it be snake_case deliberately but I'm not sure what I think make sense. Perhaps camelCase after all? I guess hub.services.<name>.apiToken and hub.config.Jupyterhub.services.0.api_token is what I think would make sense to reference with for consistency. So yeah I'll go for apiToken in this case.

So, changing to apiToken!

Reply:
Resolved by b0716dd

Reply:
Yeah, I think the scheme for putting things in the secret, exactly matching the 'path' in .Values makes good sense. That way, there's only one thing to remember.

For converting to snake_case, I think that makes sense when we are talking directly about traitlets config, but I don't think that's what we have here, since the schema is defined by the helm chart (we can talk separately about whether that's right, but assuming it's not changing here).

This PR is already great, so I think we should land it. But I think maybe a subsequent PR can maybe even more fully bake in this consistency and do away with `get_secret_value` in favor of only having `get_config` read from multiple sources. Not sure if that's feasible, but I think it would be a nice simplification!